java/vector/src/main/java/org/apache/arrow/vector/dictionary/DictionaryEncoder.java

```diff
@@ -61,43 +59,27 @@ public static ValueVector encode(ValueVector vector, Dictionary dictionary) {
     Field indexField = new Field(valueField.getName(), indexFieldType, null);
 
     // vector to hold our indices (dictionary encoded values)
-    FieldVector indices = indexField.createVector(vector.getAllocator());
-
-    // use reflection to pull out the set method
-    // TODO implement a common interface for int vectors
-    Method setter = null;
-    for (Class<?> c : Arrays.asList(int.class, long.class)) {
-      try {
-        setter = indices.getClass().getMethod("setSafe", int.class, c);
-        break;
-      } catch (NoSuchMethodException e) {
-        // ignore
-      }
+    FieldVector createdVector = indexField.createVector(vector.getAllocator());
+    if (! (createdVector instanceof BaseIntVector)) {
+      throw new IllegalArgumentException("Dictionary encoding does not have a valid int type:" +
+          createdVector.getClass());
     }
-    if (setter == null) {
-      throw new IllegalArgumentException("Dictionary encoding does not have a valid int type:" + indices.getClass());
-    }
-
-    int count = vector.getValueCount();
 
+    BaseIntVector indices = (BaseIntVector) createdVector;
     indices.allocateNew();
 
-    try {
-      for (int i = 0; i < count; i++) {
-        Object value = vector.getObject(i);
-        if (value != null) { // if it's null leave it null
-          // note: this may fail if value was not included in the dictionary
-          Object encoded = lookUps.get(value);
-          if (encoded == null) {
-            throw new IllegalArgumentException("Dictionary encoding not defined for value:" + value);
-          }
-          setter.invoke(indices, i, encoded);
+    int count = vector.getValueCount();
+
+    for (int i = 0; i < count; i++) {
+      Object value = vector.getObject(i);
+      if (value != null) { // if it's null leave it null
+        // note: this may fail if value was not included in the dictionary
+        Integer encoded = lookUps.get(value);
```

Comment:
as a follow-up to this PR you could implement a Map<Object, int> to avoid the unboxing costs.

Reply:
Thanks, good suggestion!